File: js/src/browser/navigation.js
/**
 * Navigation-related browser operations
 *
 * This module provides navigation functions that can work with or without
 * the NavigationManager for backwards compatibility.
 */

import { TIMING } from '../core/constants.js';
import { isNavigationError } from '../core/navigation-safety.js';
import { isActionStoppedError } from '../core/page-trigger-manager.js';

/**
 * Default verification function for navigation operations.
 * Verifies that navigation completed by checking:
 * - URL matches expected pattern (if provided)
 * - Page is in a ready state
 *
 * @param {Object} options - Verification options
 * @param {Object} options.page - Browser page object
 * @param {string} options.expectedUrl - Expected URL or URL pattern (optional)
 * @param {string} options.startUrl - URL before navigation
 * @returns {Promise<{verified: boolean, actualUrl: string, reason: string}>}
 */
export async function defaultNavigationVerification(options = {}) {
  const { page, expectedUrl, startUrl } = options;

  try {
    const actualUrl = page.url();

    // If expected URL is provided, verify it matches
    if (expectedUrl) {
      // Check for exact match or pattern match
      if (actualUrl === expectedUrl) {
        return { verified: true, actualUrl, reason: 'exact URL match' };
      }
      // Check if expected URL is contained in actual URL (for patterns)
      if (
        actualUrl.includes(expectedUrl) ||
        actualUrl.startsWith(expectedUrl)
      ) {
        return { verified: true, actualUrl, reason: 'URL pattern match' };
      }
      // Check if it's a regex pattern
      if (expectedUrl instanceof RegExp && expectedUrl.test(actualUrl)) {
        return { verified: true, actualUrl, reason: 'URL regex match' };
      }

      return {
        verified: false,
        actualUrl,
        reason: `URL mismatch: expected "${expectedUrl}", got "${actualUrl}"`,
      };
    }

    // No expected URL - just verify URL changed from start
    if (startUrl && actualUrl !== startUrl) {
      return { verified: true, actualUrl, reason: 'URL changed from start' };
    }

    // If no start URL and no expected URL, assume success
    return { verified: true, actualUrl, reason: 'navigation completed' };
  } catch (error) {
    if (isNavigationError(error) || isActionStoppedError(error)) {
      return {
        verified: false,
        actualUrl: '',
        reason: 'error during verification',
        navigationError: true,
      };
    }
    throw error;
  }
}

/**
 * Verify navigation operation with retry logic
 * @param {Object} options - Verification options
 * @param {Object} options.page - Browser page object
 * @param {string} options.expectedUrl - Expected URL (optional)
 * @param {string} options.startUrl - URL before navigation
 * @param {Function} options.verifyFn - Custom verification function (optional)
 * @param {number} options.timeout - Verification timeout in ms (default: TIMING.VERIFICATION_TIMEOUT)
 * @param {number} options.retryInterval - Interval between retries (default: TIMING.VERIFICATION_RETRY_INTERVAL)
 * @param {Function} options.log - Logger instance
 * @returns {Promise<{verified: boolean, actualUrl: string, reason: string, attempts: number}>}
 */
export async function verifyNavigation(options = {}) {
  const {
    page,
    expectedUrl,
    startUrl,
    verifyFn = defaultNavigationVerification,
    timeout = TIMING.VERIFICATION_TIMEOUT,
    retryInterval = TIMING.VERIFICATION_RETRY_INTERVAL,
    log = { debug: () => {} },
  } = options;

  const startTime = Date.now();
  let attempts = 0;
  let lastResult = { verified: false, actualUrl: '', reason: '' };

  while (Date.now() - startTime < timeout) {
    attempts++;
    lastResult = await verifyFn({
      page,
      expectedUrl,
      startUrl,
    });

    if (lastResult.verified) {
      log.debug(
        () =>
          `✅ Navigation verification succeeded after ${attempts} attempt(s): ${lastResult.reason}`
      );
      return { ...lastResult, attempts };
    }

    if (lastResult.navigationError) {
      log.debug(() => '⚠️  Navigation/stop detected during verification');
      return { ...lastResult, attempts };
    }

    // Wait before next retry
    await new Promise((resolve) => setTimeout(resolve, retryInterval));
  }

  log.debug(
    () =>
      `❌ Navigation verification failed after ${attempts} attempts: ${lastResult.reason}`
  );
  return { ...lastResult, attempts };
}

/**
 * Wait for URL to stabilize (no redirects happening)
 * This is a legacy polling-based approach for backwards compatibility.
 * When navigationManager is available, use waitForPageReady instead.
 *
 * @param {Object} options - Configuration options
 * @param {Object} options.page - Browser page object
 * @param {Function} options.log - Logger instance
 * @param {Function} options.wait - Wait function
 * @param {Object} options.navigationManager - NavigationManager instance (optional)
 * @param {number} options.stableChecks - Number of consecutive stable checks required (default: 3)
 * @param {number} options.checkInterval - Interval between stability checks in ms (default: 1000)
 * @param {number} options.timeout - Maximum time to wait for stabilization in ms (default: 30000)
 * @param {string} options.reason - Reason for stabilization (for logging)
 * @returns {Promise<boolean>} - True if stabilized, false if timeout
 */
export async function waitForUrlStabilization(options = {}) {
  const {
    page,
    log,
    wait,
    navigationManager,
    stableChecks = 3,
    checkInterval = 1000,
    timeout = 30000,
    reason = 'URL stabilization',
  } = options;

  // If NavigationManager is available, delegate to it
  if (navigationManager) {
    return navigationManager.waitForPageReady({ timeout, reason });
  }

  // Legacy polling-based approach
  log.debug(() => `⏳ Waiting for URL to stabilize (${reason})...`);
  let stableCount = 0;
  let lastUrl = page.url();
  const startTime = Date.now();

  while (stableCount < stableChecks) {
    // Check timeout
    if (Date.now() - startTime > timeout) {
      log.debug(
        () => `⚠️  URL stabilization timeout after ${timeout}ms (${reason})`
      );
      return false;
    }

    await wait({ ms: checkInterval, reason: 'checking URL stability' });
    const currentUrl = page.url();

    if (currentUrl === lastUrl) {
      stableCount++;
      log.debug(
        () =>
          `🔍 [VERBOSE] URL stable for ${stableCount}/${stableChecks} checks: ${currentUrl}`
      );
    } else {
      stableCount = 0;
      lastUrl = currentUrl;
      log.debug(
        () =>
          `🔍 [VERBOSE] URL changed to: ${currentUrl}, resetting stability counter`
      );
    }
  }

  log.debug(() => `✅ URL stabilized (${reason})`);
  return true;
}

/**
 * Navigate to URL with full wait for page ready
 * @param {Object} options - Configuration options
 * @param {Object} options.page - Browser page object
 * @param {Function} options.waitForUrlStabilization - URL stabilization function (legacy)
 * @param {Object} options.navigationManager - NavigationManager instance (preferred)
 * @param {Function} options.log - Logger instance (optional)
 * @param {string} options.url - URL to navigate to
 * @param {string} options.waitUntil - Wait until condition (default: 'domcontentloaded')
 * @param {boolean} options.waitForStableUrlBefore - Wait for URL to stabilize BEFORE navigation (default: true)
 * @param {boolean} options.waitForStableUrlAfter - Wait for URL to stabilize AFTER navigation (default: true)
 * @param {boolean} options.waitForNetworkIdle - Wait for all network requests to complete (default: true)
 * @param {number} options.stableChecks - Number of consecutive stable checks required (default: 3)
 * @param {number} options.checkInterval - Interval between stability checks in ms (default: 1000)
 * @param {number} options.timeout - Navigation timeout in ms (default: 240000)
 * @param {boolean} options.verify - Whether to verify the navigation (default: true)
 * @param {Function} options.verifyFn - Custom verification function (optional)
 * @param {number} options.verificationTimeout - Verification timeout in ms (default: TIMING.VERIFICATION_TIMEOUT)
 * @returns {Promise<{navigated: boolean, verified: boolean, actualUrl?: string, reason?: string}>}
 */
export async function goto(options = {}) {
  const {
    page,
    waitForUrlStabilization: stabilizeFn,
    navigationManager,
    log = { debug: () => {} },
    url,
    waitUntil = 'domcontentloaded',
    waitForStableUrlBefore = true,
    waitForStableUrlAfter = true,
    waitForNetworkIdle = true,
    stableChecks = 3,
    checkInterval = 1000,
    timeout = 240000,
    verify = true,
    verifyFn,
    verificationTimeout = TIMING.VERIFICATION_TIMEOUT,
  } = options;

  if (!url) {
    throw new Error('url is required in options');
  }

  const startUrl = page.url();

  // If NavigationManager is available, use it for full navigation handling
  if (navigationManager) {
    try {
      const navigated = await navigationManager.navigate({
        url,
        waitUntil,
        timeout,
      });

      // Verify navigation if requested
      if (verify && navigated) {
        const verificationResult = await verifyNavigation({
          page,
          expectedUrl: url,
          startUrl,
          verifyFn,
          timeout: verificationTimeout,
          log,
        });

        return {
          navigated: true,
          verified: verificationResult.verified,
          actualUrl: verificationResult.actualUrl,
          reason: verificationResult.reason,
        };
      }

      return { navigated, verified: navigated, actualUrl: page.url() };
    } catch (error) {
      if (isNavigationError(error) || isActionStoppedError(error)) {
        // Navigation was stopped by page trigger or navigation error
        // This is not a failure - it means another action took over
        return {
          navigated: false,
          verified: false,
          reason: 'navigation stopped/interrupted',
        };
      }
      throw error;
    }
  }

  // Legacy approach without NavigationManager
  try {
    // Wait for URL to stabilize BEFORE navigation (to avoid interrupting natural redirects)
    if (waitForStableUrlBefore && stabilizeFn) {
      await stabilizeFn({
        stableChecks,
        checkInterval,
        reason: 'before navigation',
      });
    }

    // Navigate to the URL
    await page.goto(url, { waitUntil, timeout });

    // Wait for URL to stabilize AFTER navigation (to ensure all redirects are complete)
    if (waitForStableUrlAfter && stabilizeFn) {
      await stabilizeFn({
        stableChecks,
        checkInterval,
        reason: 'after navigation',
      });
    }

    // Verify navigation if requested
    if (verify) {
      const verificationResult = await verifyNavigation({
        page,
        expectedUrl: url,
        startUrl,
        verifyFn,
        timeout: verificationTimeout,
        log,
      });

      return {
        navigated: true,
        verified: verificationResult.verified,
        actualUrl: verificationResult.actualUrl,
        reason: verificationResult.reason,
      };
    }

    return { navigated: true, verified: true, actualUrl: page.url() };
  } catch (error) {
    if (isNavigationError(error) || isActionStoppedError(error)) {
      console.log(
        '⚠️  Navigation was interrupted/stopped, recovering gracefully'
      );
      return {
        navigated: false,
        verified: false,
        reason: 'navigation interrupted/stopped',
      };
    }
    throw error;
  }
}

/**
 * Wait for navigation
 * @param {Object} options - Configuration options
 * @param {Object} options.page - Browser page object
 * @param {Object} options.navigationManager - NavigationManager instance (optional)
 * @param {number} options.timeout - Timeout in ms
 * @returns {Promise<boolean>} - True if navigation completed, false on error
 */
export async function waitForNavigation(options = {}) {
  const { page, navigationManager, timeout } = options;

  // If NavigationManager is available, use it
  if (navigationManager) {
    return navigationManager.waitForNavigation({ timeout });
  }

  // Legacy approach
  try {
    await page.waitForNavigation(timeout ? { timeout } : undefined);
    return true;
  } catch (error) {
    if (isNavigationError(error)) {
      console.log(
        '⚠️  waitForNavigation was interrupted, continuing gracefully'
      );
      return false;
    }
    throw error;
  }
}

/**
 * Wait for page to be fully ready (DOM loaded + network idle + no redirects)
 * This is the recommended method for ensuring page is ready for manipulation.
 *
 * @param {Object} options - Configuration options
 * @param {Object} options.page - Browser page object
 * @param {Object} options.navigationManager - NavigationManager instance (required for full functionality)
 * @param {Object} options.networkTracker - NetworkTracker instance (optional)
 * @param {Function} options.log - Logger instance
 * @param {Function} options.wait - Wait function
 * @param {number} options.timeout - Maximum time to wait (default: 30000ms)
 * @param {string} options.reason - Reason for waiting (for logging)
 * @returns {Promise<boolean>} - True if ready, false if timeout
 */
export async function waitForPageReady(options = {}) {
  const {
    page,
    navigationManager,
    networkTracker,
    log,
    wait,
    timeout = 30000,
    reason = 'page ready',
  } = options;

  // If NavigationManager is available, delegate to it
  if (navigationManager) {
    return navigationManager.waitForPageReady({ timeout, reason });
  }

  // Fallback: use network tracker directly if available
  if (networkTracker) {
    log.debug(() => `⏳ Waiting for page ready (${reason})...`);
    const startTime = Date.now();

    // Wait for network idle
    const networkIdle = await networkTracker.waitForNetworkIdle({
      timeout,
    });

    const elapsed = Date.now() - startTime;
    if (networkIdle) {
      log.debug(() => `✅ Page ready after ${elapsed}ms (${reason})`);
    } else {
      log.debug(() => `⚠️  Page ready timeout after ${elapsed}ms (${reason})`);
    }

    return networkIdle;
  }

  // Minimal fallback: just wait a bit for DOM to settle
  log.debug(() => `⏳ Waiting for page ready - minimal mode (${reason})...`);
  await wait({ ms: 1000, reason: 'page settle time' });
  return true;
}

/**
 * Wait for any ongoing navigation and network requests to complete.
 * Use this after actions that might trigger navigation (like clicks).
 *
 * @param {Object} options - Configuration options
 * @param {Object} options.page - Browser page object
 * @param {Object} options.navigationManager - NavigationManager instance
 * @param {Object} options.networkTracker - NetworkTracker instance
 * @param {Function} options.log - Logger instance
 * @param {Function} options.wait - Wait function
 * @param {number} options.navigationCheckDelay - Time to wait for potential navigation to start (default: 500ms)
 * @param {number} options.timeout - Maximum time to wait (default: 30000ms)
 * @param {string} options.reason - Reason for waiting (for logging)
 * @returns {Promise<{navigated: boolean, ready: boolean}>}
 */
export async function waitAfterAction(options = {}) {
  const {
    page,
    navigationManager,
    networkTracker,
    log,
    wait,
    navigationCheckDelay = 500,
    timeout = 30000,
    reason = 'after action',
  } = options;

  const startUrl = page.url();
  const startTime = Date.now();

  log.debug(() => `⏳ Waiting after action (${reason})...`);

  // Wait briefly for potential navigation to start
  await wait({ ms: navigationCheckDelay, reason: 'checking for navigation' });

  // Check if navigation is in progress or URL changed
  const currentUrl = page.url();
  const urlChanged = currentUrl !== startUrl;

  if (navigationManager && navigationManager.isNavigating()) {
    log.debug(() => '🔄 Navigation in progress, waiting for completion...');
    await navigationManager.waitForNavigation({
      timeout: timeout - (Date.now() - startTime),
    });
    return { navigated: true, ready: true };
  }

  if (urlChanged) {
    log.debug(() => `🔄 URL changed: ${startUrl} → ${currentUrl}`);

    // Wait for page to be fully ready
    await waitForPageReady({
      page,
      navigationManager,
      networkTracker,
      log,
      wait,
      timeout: timeout - (Date.now() - startTime),
      reason: 'after URL change',
    });

    return { navigated: true, ready: true };
  }

  // No navigation detected, just wait for network idle
  // Use shorter idle time since this is just for XHR completion, not full page load
  if (networkTracker) {
    const idle = await networkTracker.waitForNetworkIdle({
      timeout: Math.max(0, timeout - (Date.now() - startTime)),
      idleTime: 2000, // Shorter idle time for non-navigation actions
    });
    return { navigated: false, ready: idle };
  }

  return { navigated: false, ready: true };
}

File: python/src/browser_commander/browser/navigation.py
"""Navigation-related browser operations.

This module provides navigation functions that can work with or without
the NavigationManager for backwards compatibility.
"""

from __future__ import annotations

import asyncio
import time
from dataclasses import dataclass
from re import Pattern
from typing import Any, Callable

from browser_commander.core.constants import TIMING
from browser_commander.core.logger import Logger
from browser_commander.core.navigation_safety import is_navigation_error
from browser_commander.core.page_trigger_manager import is_action_stopped_error


@dataclass
class NavigationVerificationResult:
    """Result of navigation verification."""

    verified: bool
    actual_url: str
    reason: str
    navigation_error: bool = False
    attempts: int = 0


@dataclass
class GotoResult:
    """Result of goto operation."""

    navigated: bool
    verified: bool
    actual_url: str = ""
    reason: str = ""


@dataclass
class WaitAfterActionResult:
    """Result of wait_after_action operation."""

    navigated: bool
    ready: bool


async def default_navigation_verification(
    page: Any,
    expected_url: str | Pattern | None = None,
    start_url: str | None = None,
) -> NavigationVerificationResult:
    """Default verification function for navigation operations.

    Verifies that navigation completed by checking:
    - URL matches expected pattern (if provided)
    - Page is in a ready state

    Args:
        page: Browser page object
        expected_url: Expected URL or URL pattern (optional)
        start_url: URL before navigation

    Returns:
        NavigationVerificationResult with verification status
    """
    try:
        # Get current URL
        if hasattr(page, "url"):
            actual_url = page.url() if callable(page.url) else page.url
        elif hasattr(page, "current_url"):
            actual_url = page.current_url
        else:
            actual_url = ""

        # If expected URL is provided, verify it matches
        if expected_url:
            # Check for exact match
            if actual_url == expected_url:
                return NavigationVerificationResult(
                    verified=True,
                    actual_url=actual_url,
                    reason="exact URL match",
                )

            # Check if expected URL is contained in actual URL (for patterns)
            if isinstance(expected_url, str) and (
                expected_url in actual_url or actual_url.startswith(expected_url)
            ):
                return NavigationVerificationResult(
                    verified=True,
                    actual_url=actual_url,
                    reason="URL pattern match",
                )

            # Check if it's a regex pattern
            if isinstance(expected_url, Pattern) and expected_url.search(actual_url):
                return NavigationVerificationResult(
                    verified=True,
                    actual_url=actual_url,
                    reason="URL regex match",
                )

            return NavigationVerificationResult(
                verified=False,
                actual_url=actual_url,
                reason=f'URL mismatch: expected "{expected_url}", got "{actual_url}"',
            )

        # No expected URL - just verify URL changed from start
        if start_url and actual_url != start_url:
            return NavigationVerificationResult(
                verified=True,
                actual_url=actual_url,
                reason="URL changed from start",
            )

        # If no start URL and no expected URL, assume success
        return NavigationVerificationResult(
            verified=True,
            actual_url=actual_url,
            reason="navigation completed",
        )

    except Exception as error:
        if is_navigation_error(error) or is_action_stopped_error(error):
            return NavigationVerificationResult(
                verified=False,
                actual_url="",
                reason="error during verification",
                navigation_error=True,
            )
        raise


async def verify_navigation(
    page: Any,
    expected_url: str | Pattern | None = None,
    start_url: str | None = None,
    verify_fn: Callable | None = None,
    timeout: int | None = None,
    retry_interval: int | None = None,
    log: Logger | None = None,
) -> NavigationVerificationResult:
    """Verify navigation operation with retry logic.

    Args:
        page: Browser page object
        expected_url: Expected URL (optional)
        start_url: URL before navigation
        verify_fn: Custom verification function (optional)
        timeout: Verification timeout in ms (default: TIMING.VERIFICATION_TIMEOUT)
        retry_interval: Interval between retries
            (default: TIMING.VERIFICATION_RETRY_INTERVAL)
        log: Logger instance

    Returns:
        NavigationVerificationResult with verification status and attempts
    """
    if timeout is None:
        timeout = TIMING.get("VERIFICATION_TIMEOUT", 5000)
    if retry_interval is None:
        retry_interval = TIMING.get("VERIFICATION_RETRY_INTERVAL", 100)
    if verify_fn is None:
        verify_fn = default_navigation_verification

    start_time = time.time()
    attempts = 0
    last_result = NavigationVerificationResult(
        verified=False,
        actual_url="",
        reason="",
    )

    timeout_seconds = timeout / 1000

    while time.time() - start_time < timeout_seconds:
        attempts += 1
        last_result = await verify_fn(page, expected_url, start_url)

        if last_result.verified:
            if log:
                log.debug(
                    lambda _a=attempts,
                    _r=last_result: f"Navigation verification succeeded after "
                    f"{_a} attempt(s): {_r.reason}"
                )
            return NavigationVerificationResult(
                verified=last_result.verified,
                actual_url=last_result.actual_url,
                reason=last_result.reason,
                attempts=attempts,
            )

        if last_result.navigation_error:
            if log:
                log.debug(lambda: "Navigation/stop detected during verification")
            return NavigationVerificationResult(
                verified=last_result.verified,
                actual_url=last_result.actual_url,
                reason=last_result.reason,
                navigation_error=True,
                attempts=attempts,
            )

        # Wait before next retry
        await asyncio.sleep(retry_interval / 1000)

    if log:
        log.debug(
            lambda: f"Navigation verification failed after {attempts} "
            f"attempts: {last_result.reason}"
        )

    return NavigationVerificationResult(
        verified=last_result.verified,
        actual_url=last_result.actual_url,
        reason=last_result.reason,
        attempts=attempts,
    )


async def wait_for_url_stabilization(
    page: Any,
    log: Logger,
    wait_fn: Callable[[int, str], Any],
    navigation_manager: Any | None = None,
    stable_checks: int = 3,
    check_interval: int = 1000,
    timeout: int = 30000,
    reason: str = "URL stabilization",
) -> bool:
    """Wait for URL to stabilize (no redirects happening).

    This is a legacy polling-based approach for backwards compatibility.
    When navigation_manager is available, use wait_for_page_ready instead.

    Args:
        page: Browser page object
        log: Logger instance
        wait_fn: Wait function (ms, reason) -> None
        navigation_manager: NavigationManager instance (optional)
        stable_checks: Number of consecutive stable checks required (default: 3)
        check_interval: Interval between stability checks in ms (default: 1000)
        timeout: Maximum time to wait for stabilization in ms (default: 30000)
        reason: Reason for stabilization (for logging)

    Returns:
        True if stabilized, False if timeout
    """
    # If NavigationManager is available, delegate to it
    if navigation_manager:
        return await navigation_manager.wait_for_page_ready(timeout, reason)

    # Legacy polling-based approach
    log.debug(lambda: f"Waiting for URL to stabilize ({reason})...")
    stable_count = 0

    # Get current URL
    if hasattr(page, "url"):
        last_url = page.url() if callable(page.url) else page.url
    elif hasattr(page, "current_url"):
        last_url = page.current_url
    else:
        last_url = ""

    start_time = time.time()
    timeout_seconds = timeout / 1000

    while stable_count < stable_checks:
        # Check timeout
        if time.time() - start_time > timeout_seconds:
            log.debug(lambda: f"URL stabilization timeout after {timeout}ms ({reason})")
            return False

        await wait_fn(check_interval, "checking URL stability")

        # Get current URL
        if hasattr(page, "url"):
            current_url = page.url() if callable(page.url) else page.url
        elif hasattr(page, "current_url"):
            current_url = page.current_url
        else:
            current_url = ""

        if current_url == last_url:
            stable_count += 1
            log.debug(
                lambda _sc=stable_count,
                _cu=current_url: f"URL stable for {_sc}/{stable_checks} checks: {_cu}"
            )
        else:
            stable_count = 0
            last_url = current_url
            log.debug(
                lambda _cu=current_url: f"URL changed to: {_cu}, resetting stability counter"
            )

    log.debug(lambda: f"URL stabilized ({reason})")
    return True


async def goto(
    page: Any,
    url: str,
    wait_for_url_stabilization_fn: Callable | None = None,
    navigation_manager: Any | None = None,
    log: Logger | None = None,
    wait_until: str = "domcontentloaded",
    wait_for_stable_url_before: bool = True,
    wait_for_stable_url_after: bool = True,
    wait_for_network_idle: bool = True,
    stable_checks: int = 3,
    check_interval: int = 1000,
    timeout: int = 240000,
    verify: bool = True,
    verify_fn: Callable | None = None,
    verification_timeout: int | None = None,
) -> GotoResult:
    """Navigate to URL with full wait for page ready.

    Args:
        page: Browser page object
        url: URL to navigate to
        wait_for_url_stabilization_fn: URL stabilization function (legacy)
        navigation_manager: NavigationManager instance (preferred)
        log: Logger instance (optional)
        wait_until: Wait until condition (default: 'domcontentloaded')
        wait_for_stable_url_before: Wait for URL to stabilize BEFORE navigation
        wait_for_stable_url_after: Wait for URL to stabilize AFTER navigation
        wait_for_network_idle: Wait for all network requests to complete
        stable_checks: Number of consecutive stable checks required
        check_interval: Interval between stability checks in ms
        timeout: Navigation timeout in ms (default: 240000)
        verify: Whether to verify the navigation (default: True)
        verify_fn: Custom verification function (optional)
        verification_timeout: Verification timeout in ms

    Returns:
        GotoResult with navigation and verification status
    """
    if not url:
        raise ValueError("url is required")

    if verification_timeout is None:
        verification_timeout = TIMING.get("VERIFICATION_TIMEOUT", 5000)

    # Create a no-op logger if none provided
    class NoOpLogger:
        def debug(self, _: Callable[[], str]) -> None:
            pass

    if log is None:
        log = NoOpLogger()  # type: ignore

    # Get start URL
    if hasattr(page, "url"):
        start_url = page.url() if callable(page.url) else page.url
    elif hasattr(page, "current_url"):
        start_url = page.current_url
    else:
        start_url = ""

    # If NavigationManager is available, use it for full navigation handling
    if navigation_manager:
        try:
            navigated = await navigation_manager.navigate(url, wait_until, timeout)

            # Verify navigation if requested
            if verify and navigated:
                verification_result = await verify_navigation(
                    page=page,
                    expected_url=url,
                    start_url=start_url,
                    verify_fn=verify_fn,
                    timeout=verification_timeout,
                    log=log,
                )

                return GotoResult(
                    navigated=True,
                    verified=verification_result.verified,
                    actual_url=verification_result.actual_url,
                    reason=verification_result.reason,
                )

            # Get current URL for result
            if hasattr(page, "url"):
                current_url = page.url() if callable(page.url) else page.url
            elif hasattr(page, "current_url"):
                current_url = page.current_url
            else:
                current_url = ""

            return GotoResult(
                navigated=navigated,
                verified=navigated,
                actual_url=current_url,
            )

        except Exception as error:
            if is_navigation_error(error) or is_action_stopped_error(error):
                # Navigation was stopped by page trigger or navigation error
                return GotoResult(
                    navigated=False,
                    verified=False,
                    reason="navigation stopped/interrupted",
                )
            raise

    # Legacy approach without NavigationManager
    try:
        # Wait for URL to stabilize BEFORE navigation
        if wait_for_stable_url_before and wait_for_url_stabilization_fn:
            await wait_for_url_stabilization_fn(
                stable_checks=stable_checks,
                check_interval=check_interval,
                reason="before navigation",
            )

        # Navigate to the URL
        if hasattr(page, "goto"):
            # Playwright
            await page.goto(url, wait_until=wait_until, timeout=timeout)
        elif hasattr(page, "get"):
            # Selenium
            page.get(url)
        else:
            raise ValueError("Unknown page type - cannot navigate")

        # Wait for URL to stabilize AFTER navigation
        if wait_for_stable_url_after and wait_for_url_stabilization_fn:
            await wait_for_url_stabilization_fn(
                stable_checks=stable_checks,
                check_interval=check_interval,
                reason="after navigation",
            )

        # Verify navigation if requested
        if verify:
            verification_result = await verify_navigation(
                page=page,
                expected_url=url,
                start_url=start_url,
                verify_fn=verify_fn,
                timeout=verification_timeout,
                log=log,
            )

            return GotoResult(
                navigated=True,
                verified=verification_result.verified,
                actual_url=verification_result.actual_url,
                reason=verification_result.reason,
            )

        # Get current URL for result
        if hasattr(page, "url"):
            current_url = page.url() if callable(page.url) else page.url
        elif hasattr(page, "current_url"):
            current_url = page.current_url
        else:
            current_url = ""

        return GotoResult(
            navigated=True,
            verified=True,
            actual_url=current_url,
        )

    except Exception as error:
        if is_navigation_error(error) or is_action_stopped_error(error):
            print("Navigation was interrupted/stopped, recovering gracefully")
            return GotoResult(
                navigated=False,
                verified=False,
                reason="navigation interrupted/stopped",
            )
        raise


async def wait_for_navigation(
    page: Any,
    navigation_manager: Any | None = None,
    timeout: int | None = None,
) -> bool:
    """Wait for navigation to complete.

    Args:
        page: Browser page object
        navigation_manager: NavigationManager instance (optional)
        timeout: Timeout in ms

    Returns:
        True if navigation completed, False on error
    """
    # If NavigationManager is available, use it
    if navigation_manager:
        return await navigation_manager.wait_for_navigation(timeout)

    # Legacy approach
    try:
        if hasattr(page, "wait_for_load_state"):
            # Playwright
            if timeout:
                await page.wait_for_load_state("load", timeout=timeout)
            else:
                await page.wait_for_load_state("load")
        # Selenium doesn't have wait_for_navigation, so just return True
        return True
    except Exception as error:
        if is_navigation_error(error):
            print("wait_for_navigation was interrupted, continuing gracefully")
            return False
        raise


async def wait_for_page_ready(
    page: Any,
    navigation_manager: Any | None = None,
    network_tracker: Any | None = None,
    log: Logger | None = None,
    wait_fn: Callable[[int, str], Any] | None = None,
    timeout: int = 30000,
    reason: str = "page ready",
) -> bool:
    """Wait for page to be fully ready (DOM loaded + network idle + no redirects).

    This is the recommended method for ensuring page is ready for manipulation.

    Args:
        page: Browser page object
        navigation_manager: NavigationManager instance (required for full functionality)
        network_tracker: NetworkTracker instance (optional)
        log: Logger instance
        wait_fn: Wait function
        timeout: Maximum time to wait (default: 30000ms)
        reason: Reason for waiting (for logging)

    Returns:
        True if ready, False if timeout
    """

    # Create a no-op logger if none provided
    class NoOpLogger:
        def debug(self, _: Callable[[], str]) -> None:
            pass

    if log is None:
        log = NoOpLogger()  # type: ignore

    # If NavigationManager is available, delegate to it
    if navigation_manager:
        return await navigation_manager.wait_for_page_ready(timeout, reason)

    # Fallback: use network tracker directly if available
    if network_tracker:
        log.debug(lambda: f"Waiting for page ready ({reason})...")
        start_time = time.time()

        # Wait for network idle
        network_idle = await network_tracker.wait_for_network_idle(timeout, 2000)

        elapsed = int((time.time() - start_time) * 1000)
        if network_idle:
            log.debug(lambda: f"Page ready after {elapsed}ms ({reason})")
        else:
            log.debug(lambda: f"Page ready timeout after {elapsed}ms ({reason})")

        return network_idle

    # Minimal fallback: just wait a bit for DOM to settle
    log.debug(lambda: f"Waiting for page ready - minimal mode ({reason})...")
    if wait_fn:
        await wait_fn(1000, "page settle time")
    else:
        await asyncio.sleep(1)
    return True


async def wait_after_action(
    page: Any,
    navigation_manager: Any | None = None,
    network_tracker: Any | None = None,
    log: Logger | None = None,
    wait_fn: Callable[[int, str], Any] | None = None,
    navigation_check_delay: int = 500,
    timeout: int = 30000,
    reason: str = "after action",
) -> WaitAfterActionResult:
    """Wait for any ongoing navigation and network requests to complete.

    Use this after actions that might trigger navigation (like clicks).

    Args:
        page: Browser page object
        navigation_manager: NavigationManager instance
        network_tracker: NetworkTracker instance
        log: Logger instance
        wait_fn: Wait function
        navigation_check_delay: Time to wait for potential navigation to start
        timeout: Maximum time to wait (default: 30000ms)
        reason: Reason for waiting (for logging)

    Returns:
        WaitAfterActionResult with navigated and ready flags
    """

    # Create a no-op logger if none provided
    class NoOpLogger:
        def debug(self, _: Callable[[], str]) -> None:
            pass

    if log is None:
        log = NoOpLogger()  # type: ignore

    # Get start URL
    if hasattr(page, "url"):
        start_url = page.url() if callable(page.url) else page.url
    elif hasattr(page, "current_url"):
        start_url = page.current_url
    else:
        start_url = ""

    start_time = time.time()

    log.debug(lambda: f"Waiting after action ({reason})...")

    # Wait briefly for potential navigation to start
    if wait_fn:
        await wait_fn(navigation_check_delay, "checking for navigation")
    else:
        await asyncio.sleep(navigation_check_delay / 1000)

    # Get current URL
    if hasattr(page, "url"):
        current_url = page.url() if callable(page.url) else page.url
    elif hasattr(page, "current_url"):
        current_url = page.current_url
    else:
        current_url = ""

    url_changed = current_url != start_url

    if navigation_manager and navigation_manager.is_navigating():
        log.debug(lambda: "Navigation in progress, waiting for completion...")
        remaining_timeout = timeout - int((time.time() - start_time) * 1000)
        await navigation_manager.wait_for_navigation(remaining_timeout)
        return WaitAfterActionResult(navigated=True, ready=True)

    if url_changed:
        log.debug(lambda: f"URL changed: {start_url} -> {current_url}")

        # Wait for page to be fully ready
        remaining_timeout = timeout - int((time.time() - start_time) * 1000)
        await wait_for_page_ready(
            page=page,
            navigation_manager=navigation_manager,
            network_tracker=network_tracker,
            log=log,
            wait_fn=wait_fn,
            timeout=remaining_timeout,
            reason="after URL change",
        )

        return WaitAfterActionResult(navigated=True, ready=True)

    # No navigation detected, just wait for network idle
    if network_tracker:
        remaining_timeout = max(0, timeout - int((time.time() - start_time) * 1000))
        idle = await network_tracker.wait_for_network_idle(
            timeout=remaining_timeout,
            idle_time=2000,  # Shorter idle time for non-navigation actions
        )
        return WaitAfterActionResult(navigated=False, ready=idle)

    return WaitAfterActionResult(navigated=False, ready=True)
